rfcs/efficient_saccades.md

```diff
@@ -0,0 +1,111 @@
+
+### Integration
+
+### Executing Goal States
+This RFC would move us down a very goal-state-heavy path. It's easy to imagine that a distant agent performing inference will produce achievable goal states at every step. This is a pretty big departure from Monty's existing behavior in which goal states are only occasionally generated and acted upon. The remainder of the time, actions are derived from model-free policies.
+
+Implementing this RFC is going to require at least some revisions to our goal-state execution framework. First, we
+only have one goal-state-attempter -- `InformedEnvironmentDataLoader.execute_jump_attempt` -- and has only one strategy. All attempts consist of an agent-pose and sensor-orientation setting actions. This RFC entails that most of goal states will have been produced by the `SalienceMapSM`, and are therefore achievable through saccades alone. At a mimimum, we should build out some system that's capable of deciding whether a saccade is sufficient or whether it should perform pose-and-orientation resets. Second, I believe there are existing plans (desires?) to move `execute_jump_attempt` out of `InformedEnvironmentDataLoader`. I'm assuming it's headed for a `MotorPolicy`, but I don't really know.
+
+Also, I'd like to make one thing clear: The `MotorSystem`/`MotorPolicy` situation is probably my biggest Monty blindspot. I'm also aware that Tristan has been actively refactoring `Motor*` code, which means I'm at least as ignorant about the future as I am the present.
+
+### Modifying Monty Classes
+The `MontyBase` or one of its subclass will need a `goal_state_selector` attribute. Which subclass? TBD.
+  
+`SalienceMapSM` is sensor module, so it can be added to `Monty`, and it will be called like any other. But we will need to implement the passing of its up new routing pathways.
+
+**Routing**
+  - Add: `SalienceMapSM` ->  `GoalStateSelector`.
+  - Add: LM/GSG -> `GoalStateSelector`.
+  - Add: `GoalStateSelector` -> `MotorSystem`.
+  - Remove: LM/GSG -> `MotorSystem`. See `MontyForEvidenceGraphMatching._pass_infos_to_motor_system`.
+  
+**Potential issue**
+  - All `MontyBase` subclasses appear to have the same set of attributes. I don't know how much code assumes that this is the case. If `MontyExperiment` classes assume a fixed set of attributes for initialization and logging, there could be a few pain points.
+
+# Open Questions
+ - Can/should we integrate model-based signals to inform segmentation or region selection?
+ - Which segmentation methods, or combination thereof, might work well for multi-object environments (where objects do not share space) and compmositional objects (where objects *do* share space). For example, depth-based segmentation will be no help in separating a logo from the mug it is on.
```

Comment:
```suggestion
 - Which segmentation methods, or combination thereof, might work well for multi-object environments (where objects do not share space) and compositional objects (where objects *do* share space). For example, depth-based segmentation will be no help in separating a logo from the mug it is on.
```

Reply:
I think for color we will probably be able to use a similar mechanism as for depth. Basically, detecting sudden shifts in the feature distribution as borders of objects.